rust/datafusion/src/logicalplan.rs

```diff
@@ -419,13 +410,50 @@ impl Expr {
         }
     }
 
+    /// Determine if this expression can produce null values
+    pub fn nullable(&self, input_schema: &Schema) -> Result<bool> {
+        match self {
+            Expr::Alias(expr, _) => expr.nullable(input_schema),
+            Expr::Column(name) => Ok(input_schema.field_with_name(name)?.is_nullable()),
+            Expr::Literal(value) => match value {
+                ScalarValue::Null => Ok(true),
+                _ => Ok(false),
+            },
+            Expr::Cast { expr, .. } => expr.nullable(input_schema),
+            Expr::ScalarFunction { .. } => Ok(true),
+            Expr::AggregateFunction { .. } => Ok(true),
+            Expr::Not(_) => Ok(false),
```

Comment:
```suggestion
            Expr::Not(expr) => expr.nullable(input_schema),
```
At least in SQL, `!` can be null, if the input is